File: sdks/python/src/zro_sdk/modules/ipc.py
"""IPC module — inter-app message routing.

Allows apps to send messages to other apps via the runtime's IPC
routing mechanism. The module registers ``__ipc:send`` for outgoing
messages and an ``__ipc:receive`` event handler for incoming messages.

Example::

    from zro_sdk.modules import IpcModule

    ipc = IpcModule()

    @ipc.on_receive("open-file")
    async def handle_open(ctx, data):
        path = data.get("path", "")
        print(f"Received request to open: {path}")
        return {"opened": True}

    app.module(ipc)
"""

from __future__ import annotations

import logging
from typing import Any, Callable, Coroutine, Optional

from ..context import AppContext
from ..module import ModuleMeta, ModuleRegistrar, ZroModule

logger = logging.getLogger("zro.ipc")

IpcHandler = Callable[[AppContext, Any], Coroutine[Any, Any, Any]]


class IpcModule(ZroModule):
    """IPC module for inter-app communication."""

    def __init__(self) -> None:
        self._handlers: dict[str, IpcHandler] = {}

    @property
    def meta(self) -> ModuleMeta:
        return ModuleMeta(
            name="ipc",
            version="0.1.0",
            description="Inter-app message routing",
        )

    def on_receive(self, channel: str) -> Callable[[IpcHandler], IpcHandler]:
        """Decorator: register a handler for incoming messages on a named channel."""

        def decorator(func: IpcHandler) -> IpcHandler:
            self._handlers[channel] = func
            return func

        return decorator

    def register(self, r: ModuleRegistrar) -> None:
        module = self

        @r.command("__ipc:send")
        async def ipc_send(ctx: AppContext, target: str, channel: str,
                           data: Any = None) -> dict[str, str]:
            """Send a message to another app via the runtime."""
            ipc_msg = {
                "source": ctx.slug,
                "target": target,
                "channel": channel,
                "data": data,
            }
            await ctx.emit("__ipc:route", ipc_msg)
            return {"status": "sent"}

        @r.on_event("__ipc:receive")
        async def ipc_receive(ctx: AppContext, source: str = "",
                              channel: str = "", data: Any = None) -> None:
            """Handle incoming IPC messages from other apps."""
            handler = module._handlers.get(channel)
            if handler:
                try:
                    await handler(ctx, data)
                    logger.debug("IPC message handled: %s from %s", channel, source)
                except Exception:
                    logger.exception(
                        "IPC handler error: channel=%s source=%s", channel, source
                    )
            else:
                logger.debug(
                    "No handler registered for IPC channel: %s (source=%s)",
                    channel,
                    source,
                )

File: sdks/nodejs/dist-test/src/modules/ipc.js
"use strict";
/**
 * IPC module — inter-app message routing.
 *
 * Allows apps to send messages to other apps via the runtime's IPC
 * routing mechanism. Registers `__ipc:send` for outgoing messages
 * and `__ipc:receive` event handler for incoming messages.
 *
 * @example
 * ```ts
 * import { IpcModule } from '@zro/sdk';
 *
 * const ipc = new IpcModule();
 * ipc.onReceive('open-file', async (ctx, data) => {
 *     console.log(`Open file: ${data.path}`);
 *     return { opened: true };
 * });
 * app.module(ipc);
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IpcModule = void 0;
class IpcModule {
    meta = {
        name: 'ipc',
        version: '0.1.0',
        description: 'Inter-app message routing',
    };
    _handlers = new Map();
    /** Register a handler for incoming messages on a named channel. */
    onReceive(channel, handler) {
        this._handlers.set(channel, handler);
        return this;
    }
    register(r) {
        const handlers = this._handlers;
        r.command('__ipc:send', async (ctx, params) => {
            const payload = {
                target: params.target,
                channel: params.channel,
                data: params.data,
            };
            const ipcMsg = {
                source: ctx.slug,
                target: payload.target,
                channel: payload.channel,
                data: payload.data,
            };
            ctx.emit('__ipc:route', ipcMsg);
            return { status: 'sent' };
        });
        r.onEvent('__ipc:receive', async (ctx, data) => {
            const payload = {
                source: data.source ?? '',
                channel: data.channel ?? '',
                data: data.data,
            };
            const handler = handlers.get(payload.channel);
            if (handler) {
                try {
                    await handler(ctx, payload.data);
                    console.error(`[ipc] Message handled: ${payload.channel} from ${payload.source}`);
                }
                catch (err) {
                    console.error(`[ipc] Handler error: channel=${payload.channel} source=${payload.source}`, err);
                }
            }
            else {
                console.error(`[ipc] No handler for channel: ${payload.channel} (source=${payload.source})`);
            }
        });
    }
}
exports.IpcModule = IpcModule;
